### DISCUSSION TITLE
requestComplete hook and preventing double requests

### DISCUSSION THREAD

**Discussion by ivan-suhorukov** [sindresorhus/ky#585] [Ideas]
Hi!

1. I'm trying to create a SPA application with ky, so i want to show a loader on the page on each request. I can show it on `beforeRequest`, but i cannot hide it on offline-error requests. When i put the hiding code in `beforeError` its just not firing. Have i any possibility to hide loader after each request? Maybe we need `requestComplete` hook?

2. Is there a method to prevent a double-request logical errors? Lets say when user resort items with slow connection i don't want to apply new resorting requests at the same time, i want to reject all requests but first. Maybe there is need of `requestId` param, so requests with the same Id cannot fire at the same time?

Thanks!

**Comment by sindresorhus** [comment]
1. I think it's easier to just do it at the Ky callsite:

```js
try {
	isShowingSpinner = true;
	await ky(…);
} finally {
	isShowingSpinner = false;
}
```

If you need it in multiple places, you can just wrap it in a reusable function.

2. What you want is debouncing: https://github.com/sindresorhus/p-debounce?tab=readme-ov-file#pdebouncepromisefunction_

**Reply by ivan-suhorukov** [reply]
1. Ok, i done that.
2. Mm, debouncing is a different thing. It prevents execution on specified time. But there is a need of stoping execution on the uncertain time, while previous request is running.

I implement this functions by my own now, but i think there is a little lack of this abilites in library :) Thanks for answering anyway.

**Reply by sindresorhus** [reply]
> It prevents execution on specified time. But there is a need of stoping execution on the uncertain time, while previous request is running.

Sounds like you didn't click the link:

> pDebounce.promise(function_)
> Execute function_ unless a previous call is still pending, in which case, return the pending promise. **Useful, for example, to avoid processing extra button clicks if the previous one is not complete.**

**Reply by ivan-suhorukov** [reply]
Oh, i click it, but i just read the debounce/throttle article, and does not RTFM :) Its good, thanks.
